zilliqa/src/consensus.rs

```diff
@@ -677,38 +679,103 @@ impl Consensus {
+
+    fn leader_for_view(&mut self, parent_hash: Hash, view: u64) -> Option<PeerId> {
+        let parent = self.get_block(&parent_hash);
+
+        let parent = match parent {
+            Ok(Some(parent)) => parent,
+            _ => {
+                warn!("parent not found during leader for view {}", view);
+                return None;
+            }
+        };
+
+        Some(parent.committee.leader(view).peer_id)
+    }
+
+    fn committee_for_view(&mut self, parent_hash: Hash, _view: u64) -> Result<Committee> {
+        let parent = self.get_block(&parent_hash);
+
+        let parent = match parent {
+            Ok(Some(parent)) => parent,
+            _ => {
+                warn!("parent not found during leader for view");
+                return Err(anyhow!("parent not found during leader for view"));
+            }
+        };
+
+        Ok(parent.committee)
     }
 
     pub fn new_view(&mut self, _: PeerId, new_view: NewView) -> Result<Option<Block>> {
+        trace!("Received new view for height: {:?}", new_view.view);
+
+        // The leader for this view should be chosen according to the parent of the highest QC
+        // What happens when there are multiple QCs with different parents?
         // if we are not the leader of the round in which the vote counts
-        if self.get_leader(new_view.view)?.public_key != self.secret_key.node_public_key() {
+        if !self.are_we_leader_for_view(new_view.qc.block_hash, new_view.view) {
             trace!(new_view.view, "skipping new view, not the leader");
             return Ok(None);
         }
         // if the vote is too old and does not count anymore
         if new_view.view < self.view {
+            trace!(new_view.view, "Received a vote which is too old for us, discarding. Our view is: {} and new_view is: {}", self.view, new_view.view);
             return Ok(None);
         }
-        let committee = self.committee()?;
+
+        // Get the committee for the qc hash (should be highest?) for this view
+        let committee = self.committee_for_view(new_view.qc.block_hash, new_view.view)?;
         // verify the sender's signature on the block hash
         let Some((index, sender)) = committee
             .iter()
             .enumerate()
             .find(|(_, v)| v.public_key == new_view.public_key)
         else {
-            debug!("ignoring new view from unknown node (buffer?)");
+            debug!("ignoring new view from unknown node (buffer?) - committee size is : {:?} hash is: {:?} high hash is: {:?}", committee.len(), new_view.qc.block_hash, self.high_qc.block_hash);
             return Ok(None);
         };
         new_view.verify(sender.public_key)?;
 
         // check if the sender's qc is higher than our high_qc or even higher than our view
         self.update_high_qc_and_view(false, new_view.qc.clone())?;
 
-        let committee_size = self.committee()?.len();
+        let committee_size = committee.len();
+
+        if committee_size > 4 {
```

Comment:
I noticed the bitvec sometimes allocated 1024 bits for some reason, I wondered if it was committee size (nothing turned up though)